### PR TITLE
Allow `Self::cmp(self, other)` as a correct impl

### LINKED ISSUE #11178: incorrect_partial_ord_impl_on_ord_type false positive: `Some(Ord::cmp(self, other))`
### Summary

Clippy flags the following as an incorrect PartialOrd impl, with a deny-by-default correctness lint:

```rust
impl PartialOrd for T {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(Ord::cmp(self, other))
    }
}
```

in favor of preferring:

```rust
impl PartialOrd for T {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
```

But there is no correctness difference between these impl, only a stylistic difference.

### Lint Name

incorrect_partial_ord_impl_on_ord_type

### Reproducer

```rust
use std::cmp::Ordering;

pub struct Struct(String);

impl Struct {
    pub fn cmp(&self) {
        println!("Struct::cmp!");
    }
}

impl Eq for Struct {}

impl PartialEq for Struct {
    fn eq(&self, other: &Self) -> bool {
        PartialEq::eq(&self.0, &other.0)
    }
}

impl Ord for Struct {
    fn cmp(&self, other: &Self) -> Ordering {
        Ord::cmp(&self.0, &other.0)
    }
}

impl PartialOrd for Struct {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(Ord::cmp(self, other))
    }
}
```

```console
error: incorrect implementation of `partial_cmp` on an `Ord` type
  --> src/main.rs:25:1
   |
25 | /  impl PartialOrd for Struct {
26 | |      fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
   | | _____________________________________________________________-
27 | ||         Some(Ord::cmp(self, other))
28 | ||     }
   | ||_____- help: change this to: `{ Some(self.cmp(other)) }`
29 | |  }
   | |__^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incorrect_partial_ord_impl_on_ord_type
   = note: `#[deny(clippy::incorrect_partial_ord_impl_on_ord_type)]` on by default
```

I consider this a false positive for this lint. If there is a compelling stylistic argument in favor of `self.cmp(other)` over `Ord::cmp(self, other)`, that should be at most a warn-by-default style lint, not a deny-by-default correctness lint.

I think neither `Some(Ord::cmp(self, other))` nor `Some(T::cmp(self, other))` should trigger a lint.

Currently on GitHub, there are [1.6k search results for "::cmp(self"](https://github.com/search?type=code&q=%22%3A%3Acmp%28self%22+language%3ARust&l=Rust) and [26k search results for "self.cmp("](https://github.com/search?type=code&q=%22self.cmp%28%22+language%3ARust&l=Rust). Clearly there is a preference for `self.cmp(other)` but that is still a significant proportion of users who prefer to write the other, especially in macro-generated code where `self.cmp(other)` might not compile (as is the case above).

```console
error[E0061]: this method takes 0 arguments but 1 argument was supplied
  --> src/main.rs:27:19
   |
27 |         Some(self.cmp(other))
   |                   ^^^ -----
   |                       |
   |                       unexpected argument of type `&Struct`
   |                       help: remove the extra argument
   |
note: method defined here
  --> src/main.rs:6:12
   |
6  |     pub fn cmp(&self) {
   |            ^^^
```

### Version

```text
rustc 1.73.0-nightly (da6b55cc5 2023-07-17)
binary: rustc
commit-hash: da6b55cc5eaf76ed6acb7dc2f7d611e32af7c9a7
commit-date: 2023-07-17
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

### PR DESCRIPTION
**Pull request by Centri3**
Fixes #11178

Also no longer checks if the method name is *just* cmp, but the path. That was an oversight on my part ^^

r? @xFrednet
(and @blyxyas too!)

changelog: [`incorrect_partial_ord_impl_on_ord_type`]: Now allows non-method calls to `cmp` like `Self::cmp(self, other)`

### CODE REVIEW

**Comment by Centri3** [comment]
This got into beta (https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=a2f01f29c980dc6da61c1ef714cc9709), so perhaps it should be backported?

@rustbot label +beta-nominated

File: clippy_lints/src/incorrect_impls.rs
Comment (blyxyas) [inline clippy_lints/src/incorrect_impls.rs:250]: This doc-comment is kinda ambiguous imo

File: clippy_utils/src/paths.rs
Comment (blyxyas) [inline clippy_utils/src/paths.rs:164]: Wouldn't [`sym::cmp`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_span/symbol/sym/constant.cmp.html) work?

Comment (Centri3) [reply clippy_utils/src/paths.rs:164]: No, take a look at this code for example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a9afbef379712a643187be77ae228d12

If it both defines a method `cmp` and also implements `Ord`, and `Ord` is not in scope (due to `#![no_implicit_prelude]`), it'll call the method instead - *not* `Ord::cmp`. Yet it still doesn't lint

*But* the suggestion is wrong and should be fixed...

**Comment by xFrednet** [comment]
Hey @blyxyas can I give you this review? (I'm visiting my hometown for the next 3 weeks. It's totally fine if not, I should have enough time :))

**Comment by blyxyas** [comment]
Skill issue.
@bors r? @blyxyas

**Comment by xFrednet** [comment]
IDK, if having you do my work so that I can have more free time is a skill issue :wink: /s

I appreciate it! This way I'll have some time to work on Marker :D

**Comment by blyxyas** [comment]
> IDK, if having you do my work so that I can have more free time is a skill issue

I mean, phrasing it like that sounds like I'm a loser \s

**Comment by xFrednet** [comment]
Meaning it was a good come back, after you destroyed me ^^

File: tests/ui/incorrect_partial_ord_impl_on_ord_type.rs
Comment (blyxyas) [inline tests/ui/incorrect_partial_ord_impl_on_ord_type.rs:152]: Today is tomorrow ig ^w^

**Review by blyxyas — COMMENTED** [review]
LGTM, just going to wait until CI is cleared to merge ^^

**Review by blyxyas — APPROVED** [review]

**Comment by blyxyas** [comment]
@xFrednet you said to give this a review, but not about merging. Do you want to do a second-check or do I just merge?

**Comment by xFrednet** [comment]
I meant that you take over the full review ^^. But I'm happy to delegate your orders or bors:

----

*Dear lord bors, I'm pleased to announce that we have two new members in our powerful alliance. After their training, we have full trust in their magic powers. You already got the letter pigeon, informing you that they received r+ rights on this magnificent repository.*

*I'm happy to pass along the r+ from princess @blyxyas (who is apparently more skilled in time management than me). During the quest to review #11188, she has slain multiple bugs. Her support @Centri3 artfully crafted multiple commits, therefore fortifying our stronghold against the gobbling.*

*Please honer their contributions, by merging their changes into the great history book called MASTER*

---

You know, I'm always not sure if these commend are funny or come off as totally deranged. :sweat_smile: Just for the protocol, the intention is to be funny, humor can be weird ¯\\\_(ツ)\_/¯

**Comment by flip1995** [comment]
No backport necessary: This lint isn't in beta yet and this fix is already synced.

**Pull request merged**
